diff --git a/gdb/nat/linux-btrace.c b/gdb/nat/linux-btrace.c
index 5715168d2f0..acec453891f 100644
--- a/gdb/nat/linux-btrace.c
+++ b/gdb/nat/linux-btrace.c
@@ -446,6 +446,17 @@ diagnose_perf_event_open_fail ()
   error (_("Failed to start recording: %s"), safe_strerror (orig_errno));
 }
 
+// SSRVODKA: fix according to https://inbox.sourceware.org/gdb-patches/33c69c1d-7153-9cb0-ae77-f2a560927ea6@trande.de/t/#mad54825c163bbf98b0ff3ce4a09e4616c5e43642
+/* Returns System page size.  */
+static long
+system_page_size ()
+{
+  long page_size = sysconf (_SC_PAGESIZE);
+  if (page_size == -1)
+    perror_with_name (_("Failed to get system page size"));
+  return page_size;
+}
+
 /* Get the linux version of a btrace_target_info.  */
 
 static linux_btrace_target_info *
@@ -492,9 +503,10 @@ linux_enable_bts (ptid_t ptid, const struct btrace_config_bts *conf)
   if (fd.get () < 0)
     diagnose_perf_event_open_fail ();
 
+  long page_size = system_page_size ();
   /* Convert the requested size in bytes to pages (rounding up).  */
-  pages = ((size_t) conf->size / PAGE_SIZE
-	   + ((conf->size % PAGE_SIZE) == 0 ? 0 : 1));
+  pages = ((size_t) conf->size / page_size
+	   + ((conf->size % page_size) == 0 ? 0 : 1));
   /* We need at least one page.  */
   if (pages == 0)
     pages = 1;
@@ -513,17 +525,17 @@ linux_enable_bts (ptid_t ptid, const struct btrace_config_bts *conf)
       size_t length;
       __u64 data_size;
 
-      data_size = (__u64) pages * PAGE_SIZE;
+      data_size = (__u64) pages * (__u64) page_size;
 
       /* Don't ask for more than we can represent in the configuration.  */
       if ((__u64) UINT_MAX < data_size)
 	continue;
 
       size = (size_t) data_size;
-      length = size + PAGE_SIZE;
+      length = size + (size_t) page_size;
 
       /* Check for overflows.  */
-      if ((__u64) length != data_size + PAGE_SIZE)
+      if ((__u64) length != data_size + (__u64) page_size)
 	continue;
 
       errno = 0;
@@ -538,7 +550,7 @@ linux_enable_bts (ptid_t ptid, const struct btrace_config_bts *conf)
 
   struct perf_event_mmap_page *header = (struct perf_event_mmap_page *)
     data.get ();
-  data_offset = PAGE_SIZE;
+  data_offset = (__u64) page_size;
 
 #if defined (PERF_ATTR_SIZE_VER5)
   if (offsetof (struct perf_event_mmap_page, data_size) <= header->size)
@@ -633,7 +645,8 @@ linux_enable_pt (ptid_t ptid, const struct btrace_config_pt *conf)
     diagnose_perf_event_open_fail ();
 
   /* Allocate the configuration page. */
-  scoped_mmap data (nullptr, PAGE_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED,
+  long page_size = system_page_size ();
+  scoped_mmap data (nullptr, (size_t) page_size, PROT_READ | PROT_WRITE, MAP_SHARED,
 		    fd.get (), 0);
   if (data.get () == MAP_FAILED)
     error (_("Failed to map trace user page: %s."), safe_strerror (errno));
@@ -644,8 +657,8 @@ linux_enable_pt (ptid_t ptid, const struct btrace_config_pt *conf)
   header->aux_offset = header->data_offset + header->data_size;
 
   /* Convert the requested size in bytes to pages (rounding up).  */
-  pages = ((size_t) conf->size / PAGE_SIZE
-	   + ((conf->size % PAGE_SIZE) == 0 ? 0 : 1));
+  pages = ((size_t) conf->size / page_size
+	   + ((conf->size % page_size) == 0 ? 0 : 1));
   /* We need at least one page.  */
   if (pages == 0)
     pages = 1;
@@ -664,7 +677,7 @@ linux_enable_pt (ptid_t ptid, const struct btrace_config_pt *conf)
       size_t length;
       __u64 data_size;
 
-      data_size = (__u64) pages * PAGE_SIZE;
+      data_size = (__u64) pages * (__u64) page_size;
 
       /* Don't ask for more than we can represent in the configuration.  */
       if ((__u64) UINT_MAX < data_size)
@@ -735,7 +748,8 @@ linux_enable_btrace (ptid_t ptid, const struct btrace_config *conf)
 static void
 linux_disable_bts (struct linux_btrace_target_info *tinfo)
 {
-  munmap ((void *) tinfo->header, tinfo->pev.size + PAGE_SIZE);
+  long page_size = system_page_size ();
+  munmap ((void *) tinfo->header, (size_t) (tinfo->pev.size + page_size));
   close (tinfo->file);
 }
 
@@ -744,8 +758,9 @@ linux_disable_bts (struct linux_btrace_target_info *tinfo)
 static void
 linux_disable_pt (struct linux_btrace_target_info *tinfo)
 {
+  long page_size = system_page_size ();
   munmap ((void *) tinfo->pev.mem, tinfo->pev.size);
-  munmap ((void *) tinfo->header, PAGE_SIZE);
+  munmap ((void *) tinfo->header, (size_t) page_size);
   close (tinfo->file);
 }
 
diff --git a/gdbserver/linux-amd64-ipa.cc b/gdbserver/linux-amd64-ipa.cc
index a6346750f49..26a5c01ff62 100644
--- a/gdbserver/linux-amd64-ipa.cc
+++ b/gdbserver/linux-amd64-ipa.cc
@@ -18,6 +18,9 @@
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.  */
 
+#ifdef __OPENHARMONY__
+#include <asm/mman.h>
+#endif
 #include <sys/mman.h>
 #include "tracepoint.h"
 #include "linux-x86-tdesc.h"
